src/WebJobs.Script/Binding/Http/HttpRequestManager.cs

```diff
@@ -0,0 +1,151 @@
+
+        protected TraceWriter TraceWriter { get; }
+
+        public async Task<HttpResponseMessage> ProcessRequestAsync(HttpRequestMessage request, Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> processRequest, CancellationToken cancellationToken)
+        {
+            if (ThrottleExceeded() || RejectAllRequests())
+            {
+                // request cannot be queued or processed at this time,
+                return RejectRequest(request);
+            }
+
+            if (_requestQueue != null)
+            {
+                // enqueue the workitem
+                var item = new HttpRequestItem
+                {
+                    Request = request,
+                    ProcessRequest = processRequest,
+                    CancellationToken = cancellationToken,
+                    CompletionSource = new TaskCompletionSource<HttpResponseMessage>()
+                };
+                await _requestQueue.SendAsync(item, cancellationToken);
+
+                return await item.CompletionSource.Task;
+            }
+            else
+            {
+                // queue is not enabled, so just dispatch the request directly
+                return await processRequest(request, cancellationToken);
+            }
+        }
+
+        /// <summary>
+        /// For a request that will be rejected due to load, max queue length
+        /// exceeded, etc. this method will be called, allowing the
+        /// status code, headers, etc. for the request to be configured.
+        /// </summary>
+        /// <param name="request">The request to reject.</param>
+        /// <returns>The response to return.</returns>
+        protected virtual HttpResponseMessage RejectRequest(HttpRequestMessage request)
+        {
+            return new HttpResponseMessage((HttpStatusCode)429);
+        }
+
+        /// <summary>
+        /// Returns a value indicating whether all incoming requests
+        /// should be rejected, for example due to host overload, etc.
+        /// </summary>
+        /// <returns>True if requests shoudl be rejected.</returns>
+        protected virtual bool RejectAllRequests()
+        {
+            return false;
+        }
+
+        private void InitializeRequestQueue()
+        {
+            var options = new ExecutionDataflowBlockOptions
+            {
+                MaxDegreeOfParallelism = Config.MaxDegreeOfParallelism
+            };
+
+            _requestQueue = new ActionBlock<HttpRequestItem>(async item =>
+            {
+                try
+                {
+                    var response = await item.ProcessRequest(item.Request, item.CancellationToken);
+                    item.CompletionSource.SetResult(response);
+                }
+                catch (Exception ex)
+                {
+                    item.CompletionSource.SetException(ex);
+                }
+
+            }, options);
+        }
+
+        private bool ThrottleExceeded()
```

Comment:
Well, the idea here would be that more throttles may be added over time, so I didn't want it to be specific to queue length.